add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/MapViewModel.kt

```diff
@@ -148,28 +162,29 @@ class MapViewModel(application: Application) : AndroidViewModel(application) {
                 maximumResults = 1
             ).onSuccess { result ->
                 (result.geoElements.firstOrNull() as? DynamicEntityObservation)?.let { observation ->
-                    // Get the dynamic entity from the observation.
-                    observation.dynamicEntity?.let { dynamicEntity ->
-                        // Build a string for observation attributes.
-                        val stringBuilder = StringBuilder()
-                        for (attribute in arrayOf("VesselName", "CallSign", "COG", "SOG")) {
-                            val value = dynamicEntity.attributes[attribute]?.toString()
-                            // Account for when an attribute has an empty value.
-                            if (!value.isNullOrEmpty()) {
-                                stringBuilder.appendLine("$attribute: $value")
-                            }
+                    // Set the identified dynamic entity, used to display the callout.
+                    selectedGeoElement = observation.dynamicEntity
+                    // Define a new CoroutineScope to collect observation events on.
+                    observationsJob = launch(Dispatchers.IO) {
+                        // Collect observation events and update the observation string accordingly.
+                        observation.dynamicEntity?.dynamicEntityChangedEvent?.collect { dynamicEntityChangedInfo ->
+                            // Parse the observation attributes, filter out empty values, and remove
+                            // starting and ending {}s.
+                            observationString =
+                                dynamicEntityChangedInfo.receivedObservation?.attributes?.filter {
+                                    it.value.toString().isNotEmpty()
+                                }.toString().replaceFirst("{", " ").removeSuffix("}")
+                                    .replace(",", "\n")
                         }
-                        val entityAttributes = stringBuilder.toString().trimEnd()
-                        messageDialogVM.showMessageDialog(
-                            title = "Identified dynamic entity",
-                            description = entityAttributes
-                        )
                     }
+                    // If no observation is found, set the selectedGeoElement to null.
+                } ?: run {
+                    selectedGeoElement = null
```

Comment:
There can be a bug in the sample when an entity is selected after the connection is disconnected. It will display the last `observationString` which will be the same for any entity that can be identified. We should consider either not displaying a Callout when `observationString` is empty OR display a relevant message when no observations are found. 
```suggestion
                    selectedGeoElement = null
                    observationString = "Waiting for a new observation ..."
```